File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-1.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        node_hash = {None: None}
        cur = head
        while cur:
            node_hash[cur] = Node(cur.val)
            cur = cur.next
        cur = head
        while cur:
            copy = node_hash[cur]
            copy.next = node_hash[cur.next]
            copy.random = node_hash[cur.random]
            cur = cur.next
        return node_hash[head]

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode cur = head;
        int n = 0;
        while (cur != null) {
            cur = cur.next;
            n++;
        }
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        cur = dummy;
        while (n >= k) {
            n -= k;
            ListNode pre = null;
            ListNode groupHead = cur.next;
            ListNode node = groupHead;
            int i = 0;
            while (i < k) {
                i++;
                ListNode nxt = node.next;
                node.next = pre;
                pre = node;
                node = nxt;
            }
            cur.next = pre;
            groupHead.next = node;
            cur = groupHead;
        }
        return dummy.next;
    }
    
}
